pkg/controller/nstemplateset/nstemplateset_controller.go

```diff
@@ -106,6 +97,46 @@ func (r *ReconcileNSTemplateSet) Reconcile(request reconcile.Request) (reconcile
 		reqLogger.Error(err, "failed to get NSTemplateSet")
 		return reconcile.Result{}, err
 	}
+	if util.IsBeingDeleted(nsTmplSet) {
+		// if the NSTemplateSet has no finalizer, then we don't have anything to do
+		if len(nsTmplSet.Finalizers) == 0 {
+			reqLogger.Info("NSTemplateSet resource is terminated")
+			return reconcile.Result{}, nil
+		}
+		// since the NSTmplSet resource is being deleted, we must set its status to `ready=false/reason=terminating`
+		err := r.setStatusTerminating(nsTmplSet)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to set status to 'ready=false/reason=terminating' on NSTemplateSet")
+		}
+		// now, we can delete all "child" namespaces explicitly
+		username := nsTmplSet.GetName()
+		userNamespaces, err := r.fetchUserNamespaces(username)
+		if err != nil {
+			return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to list namespace with label owner '%s'", username)
+		}
+		// delete the first namespace which (still) exists and is not in a terminating state
+		reqLogger.Info("checking user namepaces associated with the deleted NSTemplateSet...")
+		for _, userNS := range userNamespaces {
+			if !util.IsBeingDeleted(&userNS) {
+				reqLogger.Info("deleting a user namepace associated with the deleted NSTemplateSet", "namespace", userNS.Name)
+				if err := r.client.Delete(context.TODO(), &userNS); err != nil {
+					return reconcile.Result{}, r.wrapErrorWithStatusUpdate(reqLogger, nsTmplSet, r.setStatusTerminatingFailed, err, "failed to delete user namespace '%s'", userNS.Name)
+				}
+				return reconcile.Result{}, nil
+			}
+		}
```

Comment:
could you please make the deletion in a separate method to make the `Reconcile` method shorter and easier to read? Thanks :-)

Reply:
sure! I wanted to do it but for some reason it slipped out of my mind 🤦‍♂

Reply:
done in b1f9439